### ISSUE TITLE
BUG - @Observable  Macro   with @attached(accessor, names: named(didSet)) 

### ISSUE THREAD

**Issue by DropDev** [apple/swift#68025]
@attached(accessor, names: named(didSet))
public macro AutoSave() = #externalMacro(module: "MyMacros", type: "AutoSaveMacro")

public struct AutoSaveMacro {}

extension AutoSaveMacro : AccessorMacro {
  public static func expansion<Context, Declaration>(of node: AttributeSyntax, providingAccessorsOf declaration: Declaration, in context: Context) throws -> [AccessorDeclSyntax] where Context : MacroExpansionContext, Declaration : DeclSyntaxProtocol {
    guard let varDecl = declaration.as(VariableDeclSyntax.self),
      let binding = varDecl.bindings.first,
      let identifier = binding.pattern.as(IdentifierPatternSyntax.self)?.identifier
    else {
      return []
    }

   return [
      """
      didSet {
        if isLoading == false {
            save()
        }
      }
      """,
}

@Observable
class Book : Codable {

var isLoading : Bool = false

func save() {
 // save
}
    let id: Int
    let title: String
   @AutoSave
    var subtitle: String
   @AutoSave
    var author: String
}

now produces either error:
1) Expansion of macro 'AutoSave()' produced an unexpected 'init' accessor
2)  Expansion of macro 'AutoSave()' did not produce a observing accessor

Xcode 15 beta 5 works and complies as expected, Beta 6 breaks this.
using @Observable seems to be the culprit in this bug.

interestingly, if you add @ObservationIgnored above @AutoSave the error goes away, but this means it can no longer be tracked. This is not the intended results. As I said this worked with being tracked in beta 5 just fine.

it should have added this:
@Observable
class Book : Codable {

var isLoading : Bool = false

func save() {
 // save
}
    let id: Int
    let title: String
   @ AutoSave
    var subtitle: String 
      didSet {
        if isLoading == false {
            save()
        }
      }
   @ AutoSave
    var author: String 
      didSet {
        if isLoading == false {
            save()
        }
      }
}

**Comment by DropDev** [comment]
Yes that Is a possibly, that something has been changed that now conflicts with the ordering, if so it requires reverting as this isn't the intended results as if you using the @Observable macro it's not then possible to use macros on it. it's also to be noted that you cannot create a peerMarco for an extra observing accessor because the @Observable macro already makes one with a suffix "_". I have tried this and it throws an error. Following this, the "_" variable is generated by the @Observable macro you cannot add the above macro to it, in the editor. It's overt there is an issue that needs addressing.

I found a previous version of Xcode 15 beta 5 in my downloads, I have completely done a full deletion of Xcode and it's data, I have reverted back to it as it's impossible to compile it on beta 6 and I can reaffirm and re-conform that the same code that does not compile in beta 6, compiles and runs properly as intended on beta 5.

**Comment by DougGregor** [comment]
I'm having a hard time understanding how this ever worked. The `@Observable` macro makes the property it is attached to into a computed property. If the `@Observable` macro sees a `didSet`/`willSet` in the original source, those can be copied over to the backing stored property (e.g., `_subtitle`). However, if the `didSet/`willSet` implementations come from a macro like `@AutoSave`, the `@Observable` macro can't see them. So while the diagnostics are in error here, I don't see how this would ever work.

**Comment by phausler** [comment]
if autosave evaluated first then it would add the `didSet` property observer accessor, which then when the observable macro would apply (post facto) it would then replicate that didSet into the storage.

But that would mean that inner macro applications would happen before outer ones; which perhaps that is what Doug is stating that won't really ever work?

**Comment by DougGregor** [comment]
> if autosave evaluated first then it would add the `didSet` property observer accessor, which then when the observable macro would apply (post facto) it would then replicate that didSet into the storage.

That's not how macros work, and they have never worked this way. The `@Observable` macro implementation sees exactly the source code as written... so it sees the `@Autosave`, but not the effects of expanding the `@Autosave` macro. I suspect this never actually did the "autosave" functionality properly.

**Comment by DropDev** [comment]
If i put a NSLog function inside the didset or save function this example calls, you can infact see that it does call those functions / logs. 

The only time it has not worked is in beta 6. 

One thing i would like to mention is that when i first worked on this on beta 5 i was running into the same error, however i did not have named(didSet) in the @attached under param names, once i read up on this this, and changed it, it made it work as intended. Then beta 6 broke this. Is maybe the named(didSet) broken in beta 6?
